aws/resource_aws_s3_object_copy.go

```diff
@@ -0,0 +1,639 @@
+		"copy_if_match",
+		"copy_if_modified_since",
+		"copy_if_none_match",
+		"copy_if_unmodified_since",
+		"customer_algorithm",
+		"customer_key",
+		"customer_key_md5",
+		"expected_bucket_owner",
+		"expected_source_bucket_owner",
+		"expires",
+		"grant_full_control",
+		"grant_read",
+		"grant_read_acp",
+		"grant_write_acp",
+		"key",
+		"kms_encryption_context",
+		"kms_key_id",
+		"metadata",
+		"metadata_directive",
+		"object_lock_legal_hold_status",
+		"object_lock_mode",
+		"object_lock_retain_until_date",
+		"request_payer",
+		"server_side_encryption",
+		"source",
+		"source_customer_algorithm",
+		"source_customer_key",
+		"source_customer_key_md5",
+		"storage_class",
+		"tagging_directive",
+		"tags",
+		"website_redirect",
+	}
+	if d.HasChanges(args...) {
+		return resourceAwsS3ObjectCopyCopy(d, meta)
+	}
+
+	return nil
+}
+
+func resourceAwsS3ObjectCopyCopy(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	input := &s3.CopyObjectInput{
+		Bucket:     aws.String(d.Get("bucket").(string)),
+		Key:        aws.String(d.Get("key").(string)),
+		CopySource: aws.String(url.QueryEscape(d.Get("source").(string))),
+	}
+
+	if v, ok := d.GetOk("acl"); ok {
+		input.ACL = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("cache_control"); ok {
+		input.CacheControl = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("content_disposition"); ok {
+		input.ContentDisposition = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("content_encoding"); ok {
+		input.ContentEncoding = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("content_language"); ok {
+		input.ContentLanguage = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("content_type"); ok {
+		input.ContentType = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("copy_if_match"); ok {
+		input.CopySourceIfMatch = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("copy_if_modified_since"); ok {
+		input.CopySourceIfModifiedSince = expandS3ObjectDate(v.(string))
+	}
+
+	if v, ok := d.GetOk("copy_if_none_match"); ok {
+		input.CopySourceIfNoneMatch = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("copy_if_unmodified_since"); ok {
+		input.CopySourceIfUnmodifiedSince = expandS3ObjectDate(v.(string))
+	}
+
+	if v, ok := d.GetOk("customer_algorithm"); ok {
+		input.SSECustomerAlgorithm = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("customer_key"); ok {
+		input.SSECustomerKey = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("customer_key_md5"); ok {
+		input.SSECustomerKeyMD5 = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("expected_bucket_owner"); ok {
+		input.ExpectedBucketOwner = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("expected_source_bucket_owner"); ok {
+		input.ExpectedSourceBucketOwner = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("expires"); ok {
+		input.Expires = expandS3ObjectDate(v.(string))
+	}
+
+	if err := resourceAwsS3ObjectCopyIncludeGrants(input, d); err != nil {
+		return err
```

Comment:
The error should be wrapped with a message that gives additional context about the error